backends/functional/test_generic.cc

```diff
@@ -0,0 +1,154 @@
+				if(it->second != state[addr]) goto error;
+			} else {
+				if(state.count_range(addr, addr + 1) != 0) goto error;
+			}
+		}
+		return;
+		error:
+		printf("FAIL\n");
+		int digits = (data_width + 3) / 4;
+		
+		for(auto addr = 0; addr < (1<<addr_width); addr++) {
+			if(addr % 8 == 0) printf("%.8x   ", addr);
+			auto it = reference.find(addr);
+			bool ref_def = it != reference.end();
+			RTLIL::Const ref_value = ref_def ? it->second : state.default_value();
+			std::string ref_string = stringf("%.*x", digits, ref_value.as_int());
+			bool sta_def = state.count_range(addr, addr + 1) == 1;
+			RTLIL::Const sta_value = state[addr];
+			std::string sta_string = stringf("%.*x", digits, sta_value.as_int());
+			if(ref_def && sta_def) {
+				if(ref_value == sta_value) printf("%s%s", ref_string.c_str(), string(digits, ' ').c_str());
+				else printf("%s%s", ref_string.c_str(), sta_string.c_str());
+			} else if(ref_def) {
+				printf("%s%s", ref_string.c_str(), string(digits, 'M').c_str());
+			} else if(sta_def) {
+				printf("%s%s", sta_string.c_str(), string(digits, 'X').c_str());
+			} else {
+				printf("%s", string(2*digits, ' ').c_str());
+			}
+			printf(" ");
+			if(addr % 8 == 7) printf("\n");
+		}
+		printf("\n");
+		//log_abort();
+	}
+	void clear_range(addr_t begin_addr, addr_t end_addr) {
+		for(auto addr = begin_addr; addr != end_addr; addr++)
+			reference.erase(addr);
+		state.clear_range(begin_addr, end_addr);
+		check();
+	}
+	void insert_concatenated(addr_t addr, RTLIL::Const const &values) {
+		addr_t words = ((addr_t) values.size() + data_width - 1) / data_width;
+		for(addr_t i = 0; i < words; i++) {
+			reference.erase(addr + i);
+			reference.emplace(addr + i, values.extract(i * data_width, data_width));
+		}
+		state.insert_concatenated(addr, values);
+		check();
+	}
+	template<typename Rnd> void run(Rnd &rnd, int n) {
+		std::uniform_int_distribution<addr_t> addr_dist(0, (1<<addr_width) - 1);
+		std::poisson_distribution<addr_t> length_dist(10);
+		std::uniform_int_distribution<uint64_t> data_dist(0, ((uint64_t)1<<data_width) - 1);
+		while(n-- > 0) {
+			addr_t low = addr_dist(rnd);
+			//addr_t length = std::min((1<<addr_width) - low, length_dist(rnd));
+			//addr_t high = low + length - 1;
+			addr_t high = addr_dist(rnd);
+			if(low > high) std::swap(low, high);
+			if((rnd() & 7) == 0) {
+				log_debug("clear %.2x to %.2x\n", (int)low, (int)high);
+				clear_range(low, high + 1);
+			} else {
+				log_debug("insert %.2x to %.2x\n", (int)low, (int)high);
+				RTLIL::Const values;
+				for(addr_t addr = low; addr <= high; addr++) {
+					RTLIL::Const word(data_dist(rnd), data_width);
+					values.bits.insert(values.bits.end(), word.bits.begin(), word.bits.end());
+				}
+				insert_concatenated(low, values);
+			}
+		}
+	}
+
+};
+
+struct FunctionalTestGeneric : public Pass
+{
+	FunctionalTestGeneric() : Pass("test_generic", "test the generic compute graph") {}
```

Comment:
I suggest we rename this to `test_cgraph` or similar. This has the potential to confuse people

Reply:
yea i'm happy to rename this, also `test_generic` needs some cleanup that i forgot to do...